plugins/army-controller-sanity.cpp

```diff
@@ -0,0 +1,133 @@
+                if (ac_set.count(ac) == 0) {
+                    WARN(log).print("acValidationError: Bad controller %p found in entity id %d\n", ac, ent->id);
+                    ok = false;
+                }
+            }
+        }
+
+        for (auto ar : df::global::world->armies.all)
+        {
+            auto ac = ar->controller;
+            if (ac && ac_set.count(ac) == 0) {
+                WARN(log).print("acValidationError: Bad controller %p found in army id %d\n", ac, ar->id);
+                ok = false;
+            }
+            else if (ac && ac->id != ar->controller_id)
+            {
+                WARN(log).print("acValidationError: controller %p id mismatch (%d != %d) in army %d\n", ac, ar->controller_id, ac->id, ar->id);
+                ok = false;
+            }
+            else if (!ac && ar->controller_id != -1)
+            {
+                WARN(log).print("acValidationError: army %d has nonzero controller %d but controller pointer is null\n", ar->id, ar->controller_id);
+                ok = false;
+            }
+        }
+
+        for (auto un : df::global::world->units.all)
+        {
+            auto ac = un->enemy.army_controller;
+            if (ac && ac_set.count(ac) == 0) {
+                WARN(log).print("acValidationError: Bad controller %p found in unit id %d\n", ac, un->id);
+                ok = false;
+            }
+            else if (ac && ac->id != un->enemy.army_controller_id)
+            {
+                WARN(log).print("acValidationError: controller %p id mismatch (%d != %d) in unit %d\n", ac, un->enemy.army_controller_id, ac->id, un->id);
+                ok = false;
+            }
+            else if (!ac && un->enemy.army_controller_id != -1)
+            {
+                WARN(log).print("acValidationError: unit %d has has nonzero controller %d but controller pointer is null\n", un->id, un->enemy.army_controller_id);
+                ok = false;
+            }
+        }
+
+        last_army_controller_next_id = *df::global::army_controller_next_id;
+        last_ac_vec_size = df::global::world->army_controllers.all.size();
+
+        INFO(log).print("acValidation: controller count = %ld, next id = %d, season tick count = %d\n",
+            last_ac_vec_size, last_army_controller_next_id, *df::global::cur_year_tick);
+
+        return ok;
+    }
+}
+
+DFhackCExport DFHack::command_result plugin_init(DFHack::color_ostream& out, std::vector <DFHack::PluginCommand>& commands)
+{
+    return DFHack::CR_OK;
+}
+
+DFhackCExport DFHack::command_result plugin_enable(DFHack::color_ostream& out, bool enable)
+{
+    is_enabled = enable;
+    return DFHack::CR_OK;
+}
+
+DFhackCExport DFHack::command_result plugin_onupdate(DFHack::color_ostream& out)
+{
+    if (is_enabled)
+    {
+        bool ok = checkArmyControllerSanity();
+        if (!ok) {
+            ERR(log).print("Army controller sanity check failed! Game pause forced.\n");
+            *df::global::pause_state = true;
```

Comment:
Can simplify to `*pause_state = true` since `REQUIRE_GLOBAL` adds a `using` statement.

You can also forego the REQUIRE_GLOBAL entirely by using `World::SetPauseState`